### PR TITLE
fix: Can not sync job status correctly when upgrading from v1.5

### LINKED ISSUE #3640: Can not sync job status correctly when upgrading from v1.5
**What happened**:

When upgrading volcano from the v1.5 to later version, the already running job created by the v1.5 volcano cannot be correctly handled. 

**Anything else we need to know?**:

v1.5 changed the naming logics of pod group by adding UID into the name: #2140, and there is also another fix regarding handling the already created pod group without UID in create or update: #2400. But a similar fix does not exist in the `syncJob` function.

So, in the following part the `syncTask` won't be set to true.

https://github.com/volcano-sh/volcano/blob/1a696ee33c88a9b3f1bf3c43f272aacfc35121be/pkg/controllers/job/job_controller_actions.go#L280-L293

**Environment**:
- Volcano Version:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by QingyaFan**
v1.5 changed the naming logics of pod group by adding UID into the name: #2140, and there is also another fix regarding handling the already created pod group without UID in create or update: #2400. But a similar fix does not exist in the syncJob function.

Fixes #3640

### CODE REVIEW

**Comment by Monokaix** [comment]
Welcome! Please sign off your commit with `git commit -s`.

File: pkg/controllers/job/job_controller_actions.go [hunk 1]
```diff
@@ -152,11 +152,18 @@ func (cc *jobcontroller) killJob(jobInfo *apis.JobInfo, podRetainPhase state.Pha
 	}
 
 	// Delete PodGroup
-	pgName := job.Name + "-" + string(job.UID)
-	if err := cc.vcClient.SchedulingV1beta1().PodGroups(job.Namespace).Delete(context.TODO(), pgName, metav1.DeleteOptions{}); err != nil {
+	pgName := jobhelpers.GetRelatedPodGroupName(job)
+	pgIface := cc.vcClient.SchedulingV1beta1().PodGroups(job.Namespace)
+	if err := pgIface.Delete(context.TODO(), pgName, metav1.DeleteOptions{}); err != nil {
```
Comment (Monokaix) [inline pkg/controllers/job/job_controller_actions.go:157]: Why named Iface here?

Comment (Monokaix) [reply pkg/controllers/job/job_controller_actions.go:157]: I mean pgIface itself is bit ambiguous: )

**Comment by Monokaix** [comment]
There are many duplicate logic when get deal with legacy pg and codes are not so readable，I think we can wrap a func to get real pg and do not expose so many logic in where controller call it.

File: pkg/controllers/job/job_controller_actions.go (hunk 1 above)
Comment (QingyaFan) [reply pkg/controllers/job/job_controller_actions.go:157]: I noticed `cc.vcClient.SchedulingV1beta1().PodGroups(job.Namespace)` returns a `v1beta1.PodGroupInterface` type, so i set the variable name to pgIface(v1beta1.**P**od**G**roup**I**nter**face**), can you give me a suggestion 🙇🏼

Comment (Monokaix) [reply pkg/controllers/job/job_controller_actions.go:157]: I think pgClient is ok.

**Comment by QingyaFan** [comment]
> There are many duplicate logic when dealing with legacy pg and codes are not so readable，I think we can wrap a func to get real pg and do not expose so many logic in where controller call it.

Yeah, that make sense. What you mean is wrap a func that check which pg(normal or legacy) exist in k8s cluster, then return it. However, this will cause a problem that: in race condition, when we get the pg name, and before we operate it, perhaps other routine has already operate it, for example maybe the pg has already been deleted. The get and the other operation are not in a transaction.

**Comment by Monokaix** [comment]
> > There are many duplicate logic when dealing with legacy pg and codes are not so readable，I think we can wrap a func to get real pg and do not expose so many logic in where controller call it.
> 
> Yeah, that make sense. What you mean is wrap a func that check which pg(normal or legacy) exist in k8s cluster, then return it. However, this will cause a problem that: in race condition, when we get the pg name, and before we operate it, perhaps other routine has already operate it, for example maybe the pg has already been deleted. The get and the other operation are not in a transaction.

That's a good point, but seems that current code is also executed in order and no lock is added，right?

**Comment by Monokaix** [comment]
> > There are many duplicate logic when dealing with legacy pg and codes are not so readable，I think we can wrap a func to get real pg and do not expose so many logic in where controller call it.
> 
> Yeah, that make sense. What you mean is wrap a func that check which pg(normal or legacy) exist in k8s cluster, then return it. However, this will cause a problem that: in race condition, when we get the pg name, and before we operate it, perhaps other routine has already operate it, for example maybe the pg has already been deleted. The get and the other operation are not in a transaction.

And if we get a pg and then call Delete method to delete the pg, if the pg is alredy deleted then NotFoundErr will also returned and we can be aware of it.

**Comment by QingyaFan** [comment]
> > > There are many duplicate logic when dealing with legacy pg and codes are not so readable，I think we can wrap a func to get real pg and do not expose so many logic in where controller call it.
> > 
> > 
> > Yeah, that make sense. What you mean is wrap a func that check which pg(normal or legacy) exist in k8s cluster, then return it. However, this will cause a problem that: in race condition, when we get the pg name, and before we operate it, perhaps other routine has already operate it, for example maybe the pg has already been deleted. The get and the other operation are not in a transaction.
> 
> And if we get a pg and then call Delete method to delete the pg, if the pg is alredy deleted then NotFoundErr will also returned and we can be aware of it.

ok, I updated the code, please review again.

File: pkg/controllers/job/job_controller_actions.go [hunk 2]
```diff
@@ -42,6 +42,30 @@ import (
 
 var calMutex sync.Mutex
 
+// getRelatedPodGroup returns the podgroup related to the vcjob.
+// it will return normal pg if it exist in cluster,
+// else it return legacy pg before version 1.5.
+func (cc *jobcontroller) getRelatedPodGroup(job *batch.Job) (*scheduling.PodGroup, error) {
```
Comment (Monokaix) [inline pkg/controllers/job/job_controller_actions.go:48]: getRelatedPodGroup -》 getPodGroupByJob

File: pkg/controllers/job/job_controller_actions.go [hunk 3]
```diff
@@ -42,6 +42,30 @@ import (
...
+	pgName := cc.generateRelatedPodGroupName(job)
```
Comment (Monokaix) [inline pkg/controllers/job/job_controller_actions.go:49]:
```go
	pgName := cc.generateRelatedPodGroupName(job)
	pg, err := cc.pgLister.PodGroups(job.Namespace).Get(pgName)
	if err == nil {
		return pg, nil
	}
	if apierrors.IsNotFound(err) {
		pg, err = cc.pgLister.PodGroups(job.Namespace).Get(job.Name)
		if err != nil {
			return nil, err
		}
		return pg, nil
	}
	return nil, err
Change to this is better: )

File: pkg/controllers/job/job_controller_actions.go [hunk 4]
```diff
@@ -152,12 +176,13 @@ func (cc *jobcontroller) killJob(jobInfo *apis.JobInfo, podRetainPhase state.Pha
 	}
 
 	// Delete PodGroup
-	pgName := job.Name + "-" + string(job.UID)
-	if err := cc.vcClient.SchedulingV1beta1().PodGroups(job.Namespace).Delete(context.TODO(), pgName, metav1.DeleteOptions{}); err != nil {
-		if !apierrors.IsNotFound(err) {
-			klog.Errorf("Failed to delete PodGroup of Job %v/%v: %v",
-				job.Namespace, job.Name, err)
-			return err
+	pg, _ := cc.getRelatedPodGroup(job)
```
Comment (Monokaix) [inline pkg/controllers/job/job_controller_actions.go:179]: We should check the returned err first.

**Comment by QingyaFan** [comment]
I improved the code according to your comment. Please take a look again :) @Monokaix

**Comment by QingyaFan** [comment]
I accidently close the PR when i rebase code just now, now I have solve the commit message problem, it's ok to continue. @Monokaix

**Comment by QingyaFan** [comment]
@Monokaix seems "E2E Sequence Jobs" failed, i have checked the output, can't find what's wrong, can you help me?

**Comment by Monokaix** [comment]
> @Monokaix seems "E2E Sequence Jobs" failed, i have checked the output, can't find what's wrong, can you help me?

I will check that later.

**Comment by Monokaix** [comment]
You can use `git commit --ame --no-edit -s` and for push again to retest.

**Comment by QingyaFan** [comment]
> New changes are detected. LGTM label has been removed.

Please help add the lgtm label again to retest @Monokaix

**Comment by william-wang** [comment]
@QingyaFan please add ut for this change.

**Comment by QingyaFan** [comment]
> @QingyaFan please add ut for this change.

ok，I will try to add ut.

**Comment by QingyaFan** [comment]
> @QingyaFan please add ut for this change.

ut has been added, please check it out when you have time.

**Comment by QingyaFan** [comment]
@Monokaix can you help me review the unit test code and merge my pull request, @william-wang seems have no time

**Comment by QingyaFan** [comment]
Please help review the unit test code and merge the pull request @Monokaix @hwdef @hzxuzhonghu @william-wang

**Comment by Monokaix** [comment]
line 205 in your pr of file `job_controller_actions_test.go` didn't check the expected err and actual err, we should modify it: )

**Comment by QingyaFan** [comment]
> line 205 in your pr of file `job_controller_actions_test.go` didn't check the expected err and actual err, we should modify it: )

however，i did not change line 205...

**Comment by Monokaix** [comment]
HI，please rebase your pr: )

**Comment by QingyaFan** [comment]
@Monokaix I fix the conflict. Can you add lgtm label. And the `Vcctl Test / E2E about Volcano CLI (pull_request)` test failed, can you retrigger the test to see if this happens occasionally(because the pr did not change the failed test file). Thanks !

**Comment by hwdef** [comment]
> ok, I updated the code, please review again.

Please squash your commit to one, the CI will restart automatically

**Comment by QingyaFan** [comment]
> > ok, I updated the code, please review again.
> 
> Please squash your commit to one, the CI will restart automatically

the commits are split in different branches and not continus, i don't know how to squash into one. what about i redo the change in a different branch and create a new pull request?

**Comment by Monokaix** [comment]
> > > ok, I updated the code, please review again.
> > 
> > 
> > Please squash your commit to one, the CI will restart automatically
> 
> the commits are split in different branches and not continus, i don't know how to squash into one. what about i redo the change in a different branch and create a new pull request?

If you are not familiar with git，you can build a new branch such as `tmp` locally based on the current branch first, and reset your current branch to latest master，and then just cherry pick your own commit from the `tmp` branch.

**Comment by Monokaix** [comment]
> > > > ok, I updated the code, please review again.
> > > 
> > > 
> > > Please squash your commit to one, the CI will restart automatically
> > 
> > 
> > the commits are split in different branches and not continus, i don't know how to squash into one. what about i redo the change in a different branch and create a new pull request?
> 
> If you are not familiar with git，you can build a new branch such as `tmp` locally based on the current branch first, and reset your current branch to latest master，and then just cherry pick your own commit from the `tmp` branch.


> > > ok, I updated the code, please review again.
> > 
> > 
> > Please squash your commit to one, the CI will restart automatically
> 
> the commits are split in different branches and not continus, i don't know how to squash into one. what about i redo the change in a different branch and create a new pull request?

You can do that: )

**Comment by QingyaFan** [comment]
> > > > > ok, I updated the code, please review again.
> > > > 
> > > > 
> > > > Please squash your commit to one, the CI will restart automatically
> > > 
> > > 
> > > the commits are split in different branches and not continus, i don't know how to squash into one. what about i redo the change in a different branch and create a new pull request?
> > 
> > 
> > If you are not familiar with git，you can build a new branch such as `tmp` locally based on the current branch first, and reset your current branch to latest master，and then just cherry pick your own commit from the `tmp` branch.
> 
> > > > ok, I updated the code, please review again.
> > > 
> > > 
> > > Please squash your commit to one, the CI will restart automatically
> > 
> > 
> > the commits are split in different branches and not continus, i don't know how to squash into one. what about i redo the change in a different branch and create a new pull request?
> 
> You can do that: )

I create a new pull request, please have a look: https://github.com/volcano-sh/volcano/pull/3786 @Monokaix

**Comment by Monokaix** [comment]
tracked by https://github.com/volcano-sh/volcano/pull/3786

**Pull request closed without merging**